Metric proto

dabacon opened the pull request.

dabacon commented:
This will fail format.  Going to add an yapf ignore in a separate pull request

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,79 @@
+// Performance metrics for a quantum processor.
+message Metrics {
+
+  // The time the metrics were collected, in milliseconds since Jan 1, 1970.
```
This quantity is ambiguous. Refer to a specific time standard, such as unix time or UTC. Include it in the field name.

dabacon replied:
Unix time is in UTC.  

I've added words to say it is unix time.  I think calling it "unix time" doesn't really help, but labeling the time unit is important.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,79 @@
+  uint64 collection_time_ms = 1;
+
+  // Map from a metric type to a collection of the metrics all of this type.
+  // Keys should be snake_case and fairly descriptive of the metric.
```
Why is the format encouraging a style guide on its keys?

dabacon replied:
because standards for casing make you not have to remember whether it was T1_time or T1Time etc.

Strilanc replied:
Oh right this is coming from the server, so we can enforce things like that.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,79 @@
+option java_multiple_files = true;
+
+// Performance metrics for a quantum processor.
+message Metrics {
```
It's confusing that a `Metrics` contains many `MetricCollection`s instead of many `Metric`s.

Maybe `NamedMetricCollections`?

dabacon replied:
Ugh.  That's like java ;)  
 How about PerformanceSnapshot?

Strilanc replied:
Yeah that's better.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,79 @@
+
+// A collection of metrics of the same type.
+message MetricCollection {
+  // The metrics in the collection. These should all be of the same type.
```
Why not have the oneof at this level if we're going to require that they all be the same type?

dabacon replied:
you can't have repeated oneofs, so if you do that you will end up making container classes for each metric type.  Do you think that's better?

Strilanc replied:
I think we could encapsulate this into the "target" of a metric, with a special all target. It will also give us flexibility to talk about subsets of qubits.

dabacon commented:
Hey look I responded faster than the label could be removed.

Strilanc commented:
You didn't respond to the table suggestion ;)

dabacon commented:
Ah missed your main comment.

I think that solution is fine as well (though we need to have collection time).  Let me refactor real quick here.

dabacon commented:
Note that you'll want repeated values for things like tomography, so I don't think you want a single "value"

dabacon commented:
PTAL.  Simplified per suggestion.  Did keep repeated values (since this makes sense for things like tomography).  Dumped float value and added string.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,51 @@
+// one of the types.
+message Value {
+  // The raw value.
+  oneof val {
```
Maybe we could have the repetition in here, instead of in `Metric`? So you can have a `Union[..., List[Float]]` instead of a `List[Union[..., Float]]`.

dabacon replied:
Repeated oneofs add a lot of indirection.  Plus it's not clear they need to be of the same type, actually.

dabacon replied:
I've deleted the comment about same type.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,51 @@
+
+    int64 int64_val = 3;
+
+    string str_value = 4;
```
val val val value

dabacon replied:
fixed.

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,51 @@
+
+// A snapshot of the performance metrics for a quantum processor at a
+// particular time.
+message MetricSnapshot {
```
`MetricsSnapshot`

Strilanc commented on cirq/api/google/v2/metrics.proto:
```diff
@@ -0,0 +1,51 @@
+  // The time the metrics were collected, in unix time (milliseconds since
+  // Epoch minus leap seconds). Metric collection take time, so this is the
+  // time at which all of the metrics have been collected.
+  uint64 collection_time_ms = 1;
```
How about `timestamp_ms`? There's not really a single well defined time when all the metrics were collected, but there can be a well defined time when we stamped it.

dabacon replied:
done

dabacon commented:
PTAL

Strilanc commented:
Seems to be missing some of the changes.

dabacon commented:
I see the changes.
